### PR TITLE
threadpool: fix race

### PR DESCRIPTION
**Pull request by BusyJay**
Notify should be called while holding the lock.

### CODE REVIEW

**Comment by BusyJay** [comment]
PTAL

File: src/util/threadpool.rs [hunk 1]
```diff
@@ -242,55 +249,52 @@ where
...
-                self.ctx.on_task_started();
-                (t.task).call_once((&mut self.ctx,));
-                self.ctx.on_task_finished();
-                self.task_count.fetch_sub(1, AtomicOrdering::SeqCst);
-                self.task_counter += 1;
-                if self.task_counter == self.tasks_per_tick {
-                    self.task_counter = 0;
-                    self.ctx.on_tick();
+        let &(ref lock, ref cvar) = &*self.state;
+        let mut state = lock.lock().unwrap();
```
Comment (siddontang) [inline src/util/threadpool.rs]: this will lock forever?

Comment (BusyJay) [reply src/util/threadpool.rs]: Release at L278 or L281.

Comment (siddontang) [reply src/util/threadpool.rs]: got it

**Comment by siddontang** [comment]
CI failed

**Comment by BusyJay** [comment]
PTAL

**Comment by siddontang** [comment]
PTAL @overvenus

**Comment by siddontang** [comment]
PTAL @zhangjinpeng1987

File: src/util/threadpool.rs [hunk 2]
```diff
@@ -242,54 +249,63 @@ where
...
         }
     }
 
-    fn get_task_timeout(&mut self, timeout: Option<Duration>) -> Option<Task<C>> {
-        let &(ref lock, ref cvar) = &*self.task_queue;
-        let mut task_queue = lock.lock().unwrap();
-
-        if let Some(task) = task_queue.pop() {
-            return Some(task);
+    fn next_task(&mut self, timeout: Duration) -> Option<Task<C>> {
```
Comment (overvenus) [inline src/util/threadpool.rs]: I think we do not need to pass timeout, it always be `NAP_SECS` and it does not return after timeout.

**Comment by disksing** [comment]
LGTM.

**Pull request merged**